File: src/mbforge/model_server/routers/moldet.py
"""MolDet 分子检测与识别路由."""

from __future__ import annotations

import asyncio
from typing import Any

from fastapi import APIRouter, Request

from ...utils.exceptions import ModelNotAvailableError, ValidationError
from ...utils.helpers import decode_base64_image
from ...utils.logger import get_logger
from ..models.moldet import get_moldet
from .health import set_model_status

logger = get_logger(__name__)
router = APIRouter()


@router.post("/detect-page")
async def detect_page(request: Request) -> dict[str, Any]:
    try:
        body = await request.json()
        image_base64 = body.get("image_base64", "")
        if not image_base64:
            raise ValidationError("image_base64 is required")

        image = decode_base64_image(image_base64)
        pipeline = get_moldet()
        if pipeline is None or not pipeline.is_available():
            raise ModelNotAvailableError("MolDet pipeline not available")

        loop = asyncio.get_running_loop()
        boxes = await loop.run_in_executor(
            None, lambda: pipeline.doc_detector.detect(image)
        )
        set_model_status("moldet", "ready")
        return {
            "boxes": [
                {"x1": x1, "y1": y1, "x2": x2, "y2": y2, "conf": conf}
                for x1, y1, x2, y2, conf in boxes
            ],
            "count": len(boxes),
        }
    except (ValidationError, ModelNotAvailableError):
        raise
    except Exception as e:
        set_model_status("moldet", "error")
        logger.error(f"MolDet detect-page failed: {e}", exc_info=True)
        raise ModelNotAvailableError(str(e))


@router.post("/extract-page")
async def extract_page(request: Request) -> dict[str, Any]:
    try:
        body = await request.json()
        image_base64 = body.get("image_base64", "")
        if not image_base64:
            raise ValidationError("image_base64 is required")

        page_idx = body.get("page_idx", 0)
        page_w_pts = body.get("page_w_pts", 595.0)
        page_h_pts = body.get("page_h_pts", 842.0)
        image_w = body.get("image_w", 0)
        image_h = body.get("image_h", 0)
        dpi = body.get("dpi", 300.0)

        if image_w == 0 or image_h == 0:
            raise ValidationError("image_w and image_h are required")

        image = decode_base64_image(image_base64)

        pipeline = get_moldet()
        if pipeline is None or not pipeline.is_available():
            raise ModelNotAvailableError("MolDet pipeline not available")

        loop = asyncio.get_running_loop()
        results = await loop.run_in_executor(
            None, lambda: pipeline.extract_page(
                image, page_idx, page_w_pts, page_h_pts, image_w, image_h, dpi
            )
        )
        set_model_status("moldet", "ready")
        return {
            "results": [r.to_dict() for r in results],
            "count": len(results),
        }
    except (ValidationError, ModelNotAvailableError):
        raise
    except Exception as e:
        set_model_status("moldet", "error")
        logger.error(f"MolDet extract-page failed: {e}", exc_info=True)
        raise ModelNotAvailableError(str(e))

File: frontend/src/api/moldet.ts
import { fetchJson } from './http'
import type { ExtractionResult } from '../types'

const API_BASE = '/api/v1/moldet'

/** 检测页面中的分子结构（仅返回 bounding boxes） */
export interface DetectPageResponse {
  boxes: { x1: number; y1: number; x2: number; y2: number; conf: number }[]
  count: number
}

export function detectPage(
  imageBase64: string,
): Promise<DetectPageResponse> {
  return fetchJson(`${API_BASE}/detect-page`, {
    method: 'POST',
    body: JSON.stringify({ image_base64: imageBase64 }),
  })
}

/** 提取页面中的分子（检测 + MolScribe 识别） */
export interface ExtractPageResponse {
  results: ExtractionResult[]
  count: number
}

export function extractPage(
  imageBase64: string,
  pageIdx: number,
  pageWPts: number,
  pageHPts: number,
  imageW: number,
  imageH: number,
): Promise<ExtractPageResponse> {
  return fetchJson(`${API_BASE}/extract-page`, {
    method: 'POST',
    body: JSON.stringify({
      image_base64: imageBase64,
      page_idx: pageIdx,
      page_w_pts: pageWPts,
      page_h_pts: pageHPts,
      image_w: imageW,
      image_h: imageH,
    }),
  })
}

/** 对已裁剪区域进行精检测 + 识别 */
export function extractRegion(
  imageBase64: string,
  pageIdx: number,
  bboxPdf: [number, number, number, number],
): Promise<{ result: ExtractionResult }> {
  return fetchJson(`${API_BASE}/extract-region`, {
    method: 'POST',
    body: JSON.stringify({
      image_base64: imageBase64,
      page_idx: pageIdx,
      bbox_pdf: bboxPdf,
    }),
  })
}
